netket/nn/symmetric_linear.py

```diff
@@ -211,35 +594,238 @@ def full_bias(self, bias):
         return jnp.repeat(bias, self.n_symm)
 
     @compact
-    def __call__(self, inputs: Array) -> Array:
+    def __call__(self, x: Array) -> Array:
         """Applies the equivariant transform to the inputs along the last dimension.
         Args:
-          inputs: The nd-array to be transformed.
+          x: The nd-array to be transformed.
         Returns:
           The transformed input.
         """
-        dtype = jnp.promote_types(inputs.dtype, self.dtype)
-        inputs = jnp.asarray(inputs, dtype)
+        dtype = jnp.promote_types(x.dtype, self.dtype)
+        x = jnp.asarray(x, dtype)
```

Comment:
OTOH I'm not sure this is needed. I'd expect `dot_general` promotes types as needed, so we might needlessly blow memory usage up by a factor of 2

Reply:
no, most `jax.lax` functions do not promote types as needed, and you are supposed to provide 2 tensors of the same type.
That's why there's the code on top.

Besides, a general limitation of XLA is that deep down you can only do operations on two tensors of the same dtype, barring few exceptions.